mesonbuild/interpreter/interpreter.py

```diff
@@ -3381,3 +3383,61 @@ def func_range(self, node, args: T.Tuple[int, T.Optional[int], T.Optional[int]],
         if step < 1:
             raise InterpreterException('step must be >=1')
         return P_OBJ.RangeHolder(start, stop, step, subproject=self.subproject)
+
+    @FeatureNew('response_file', '1.2.0')
+    @typed_pos_args('response_file', varargs=(str, mesonlib.File))
+    @typed_kwargs(
+        'response_file',
+        KwargInfo('prefix', str, default='@'),
+        KwargInfo('quote', (bool, str), default=True),  # TODO: choices...
+        KwargInfo('max_args_length', (bool, int), default=127),  # 127 is the limit specified in msvc documentation
+        KwargInfo('separator', str, default=' '),
+        KwargInfo('relative_paths', bool, default=True),
+    )
+    def func_response_file(self, node: mparser.BaseNode, args: T.Tuple[T.Union[mesonlib.FileOrString, T.List[mesonlib.FileOrString]]], kwargs: kwtypes.ResponseFile) -> T.Union[T.List[str], str]:
+
+        import shlex
+        from ..backend.ninjabackend import cmd_quote, gcc_rsp_quote
+
+        arguments = mesonlib.listify(list(args), flatten=True)
+        if not arguments:
+            return []
+
+        max_args_length = kwargs['max_args_length']
+        if max_args_length is False:
+            return arguments
+        elif max_args_length is True:
+            max_args_length = 0
+
+        quote_fcts = {
+            'sh': shlex.quote,
+            'gcc': gcc_rsp_quote,
+            'cmd': cmd_quote,
+            True: mesonlib.quote_arg,
```

Comment:
I think the important thing to note that there is a difference between gcc response file quoting (a backslash *always* escapes the following character, even inside single quotes) and sh quoting.

Since the respose file parsing is done by the invoked program, not a shell, it could plausibly be any of the alternatives, irrespective of platform.

I'm not sure how useful the "platform default quote style" might be, but it seems a reasonable thing to support, but we might say more clearly what we think it is...

Reply:
`gcc` and `sh` values are definitely needed. What I'm not convinced is if `cmd` and `True` are needed. Should `sh`, `cmd` and `True` all do `mesonlib.quote_arg`?

What's the difference in the result of those 2 functions?
- https://github.com/mesonbuild/meson/blob/master/mesonbuild/utils/universal.py#L1102
- https://github.com/mesonbuild/meson/blob/master/mesonbuild/backend/ninjabackend.py#L67

Reply:
I agree that  `mesonlib.quote_arg`  on windows and `cmd_quote` do things which should be the same (and certainly look similar).

But the platform determination made in defining `mesonlib.quote_arg` is not what's wanted in the ninja backend. There probably are some edge cases e.g. using MSVC CL on msys2 where platform isn't windows but cmd style quoting is needed in the response file.

This circles back to my point that it's the tool that determines the quoting style, not the default shell on the platform.  

Making `sh` use `mesonlib.quote_arg` seems both inconsistent in a susprising way (you've asked for `sh` but you actually get `cmd`), and perhaps disastrous on Windows, if the tool really always uses sh style quoting on all platforms (and vice versa a similiar inconsistency for `cmd` on non-Windows).

Reply:
I think this really depends on what is expected by the consumer. This could be platform dependent, or not. That's why I think it is better to have the choice.

Reply:
non-native quoting could also make sense in cross-compiling context

Reply:
I think this is correct now.